### PR TITLE
Contextual typing based on instantiated types

### LINKED ISSUE #6611: inference strangeness 6
``` typescript
export interface A<a> {
    value: a;
}

function fn<a>(values: A<a>, value: a) : void {
}

let handlers: A<(value: number) => void>;
fn(handlers, value => alert(value)); // < -- expected value: number, actual value: any

```

### LINKED ISSUE #21382: Quickinfo for contextually typed arrow function parameter can be incorrect
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Search Terms:** 

**Code**

```ts
interface BaseProps<T> {
  initialValues: T;
  nextValues: (cur: T) => T;
}
declare class Component<P> { constructor(props: P); props: P; }
declare class GenericComponent<Props = {}, Values = object>
    extends Component<Props & BaseProps<Values>> {
  iv: Values;
}

new GenericComponent({ initialValues: 12, nextValues: val => 12 });
```

**Expected behavior:**
Hover over `val` - see `number`. (Also: type `val.` in the lambda body and trigger completions, get completions for numbers). Hover over `nextValues`, see `(val: number) => number`.

**Actual behavior:**
`val`'s quick info is `any` (compilation and errors show it is assigned the correct type during compilation, the quick info on `GenericComponent` is thankfully correct), has no completions (being `any`), and `nextValues` has a quickinfo type of `(cur: number) => number & (val: number) => number`, which is an intersection of two conspicuously identical signatures.

[**Playground Link**](http://www.typescriptlang.org/play/#src=interface%20BaseProps%3CT%3E%20%7B%0D%0A%20%20initialValues%3A%20T%3B%0D%0A%20%20nextValues%3A%20(cur%3A%20T)%20%3D%3E%20T%3B%0D%0A%7D%0D%0Adeclare%20class%20Component%3CP%3E%20%7B%20constructor(props%3A%20P)%3B%20props%3A%20P%3B%20%7D%0D%0Adeclare%20class%20GenericComponent%3CProps%20%3D%20%7B%7D%2C%20Values%20%3D%20object%3E%0D%0A%20%20%20%20extends%20Component%3CProps%20%26%20BaseProps%3CValues%3E%3E%20%7B%0D%0A%20%20iv%3A%20Values%3B%0D%0A%7D%0D%0A%0D%0Anew%20GenericComponent(%7B%20initialValues%3A%2012%2C%20nextValues%3A%20val%20%3D%3E%2012%20%7D)%3B%0D%0A) (Also, [here's a fourslash test](https://gist.github.com/weswigham/76716e64ccf18957da5842ffb5067dab) to start anyone working on this off)

Preliminary investigation (read: believing this bug was related to JSX before trying the same construct outside a JSX tag) led me to believe this has something to do with execution order, as inference statefully assigns types to symbols via `assignTypeToParameterAndFixTypeParameters`, meaning if inference happens it assigns a type to the symbol, whereas if that hasn't happened and there's no cached type from an inference triggered by a `check` call, quickinfo only finds the contextual type for the position. _However_ requesting errors (which does a full typecheck) prior to checking quickinfo doesn't hide the error (which would be expected if the issue depended on execution order), so I'm unsure what the underlying problem is. The correct types are returned at these positions for our `.types` baselines by `getTypeAtLocation`, which, in these cases, is just a roundabout way to say `getTypeOfSymbol` on the node's symbol, which is the same as `getTypeOfSymbolAtLocation` in these locations (which is what's used in services).

### LINKED ISSUE #22149: Type inference for conditional callback parameters breaks down
This is based on #21879; I have a work-around for that issue but still having a separate issue.

**TypeScript Version:** 2.8.0-dev.20180204

**Code**

```ts
declare function broke(impossible: never): never; // used to ensure full case coverage

interface Foo { kind: 'foo'; }
declare function isFoo(foobar: Foo | Bar): foobar is Foo;
interface Bar { kind: 'bar'; }
declare function isBar(foobar: Foo | Bar): foobar is Bar;

function mapFooOrBar<FooBar extends Foo | Bar, R>(
    foobar: FooBar,
    mapFoo: FooBar extends Foo ? ((value: Foo) => R) : ((impossible: never) => never),
    mapBar: FooBar extends Bar ? ((value: Bar) => R) : ((impossible: never) => never),
): R {
    if (isFoo(foobar)) {
        // workaround for #21879
        const handle = mapFoo as (value: Foo) => R;
        return handle(foobar);
    }
    else if (isBar(foobar)) {
        // workaround for #21879
        const handle = mapBar as (value: Bar) => R;
        return handle(foobar);
    }
    else {
        // workaround for #20375
        return broke(foobar as never);
    }
}

declare function toFoo(): Foo;

mapFooOrBar(
    toFoo(),
    foo => foo.kind,
/*  ^^^
Parameter 'foo' implicitly has an 'any' type. */
    bar => broke(bar)
/*  ^^^^^^^^^^^^^^^^^
                 ^^^
Argument of type '(bar: any) => any' is not assignable to parameter of type '(impossible: never) => never'.
  Type 'any' is not assignable to type 'never'. */
);

mapFooOrBar<Foo, string>(
    toFoo(),
    foo => foo.kind,
    bar => broke(bar)
);
```
(the casting `foobar as never` for the `broke` function is a workaround for #20375)

**Expected Behavior**
Compile without errors.

**Actual Behavior**
Typescript fails to infer the type of `foo` or `bar` in the callbacks, despite correctly inferring the type of the generic `FooBar` parameter in `mapFooOrBar`. Hovering over the first invocation of `mapFooOrBar` in VS Code, I see
```ts
function mapFooOrBar<Foo, {}>(foobar: Foo, mapFoo: (value: Foo) => {}, mapBar: (impossible: never) => never): {}
```
Which all looks correct, excepting the `{}` instead of `string` or `'foo'` as the second parameter. But despite getting all that right, the parameters in the callback are inferred as `any`, which is likely the source of the `{}` second parameter. (Eliminating the second parameter and making the callbacks return `void` does not improve matters.)

### LINKED ISSUE #25299: Type inference issue related to usage of conditional type
**TypeScript Version:**  
2.8.1, 2.9.2, 3.0.0-dev.20180628

**Search Terms:** 
conditional type inference

**Code**

```ts
// Don't know what causes the problem so cannot come up with a shorter sample

type ActionType<P> = string & { attachPayloadTypeHack?: P & never }

/*
// irrelevant type to reproduce my issue
// but gives context that action might have no additional payload to handle

type Action<P> = P extends void
    ? { type: ActionType<P> }
    : { type: ActionType<P>, payload: P }
*/

type Handler<S, P> = P extends void
    ? (state: S) => S
    : (state: S, payload: P) => S

interface ActionHandler<S, P> {
    actionType: ActionType<P>
    handler: Handler<S, P>
}

declare function handler<S, P>(actionType: ActionType<P>, handler: Handler<S, P>): ActionHandler<S, P>

declare function createReducer<S>(
        defaultState: S,
        ...actionHandlers: ActionHandler<S /* force state S for every action handler*/, any>[]
    ): any

interface AppState {
    dummy: string
}

const defaultState: AppState = {
    dummy: ''
}

const NON_VOID_ACTION: ActionType<number> = 'NON_VOID_ACTION'
    , VOID_ACTION: ActionType<void> = 'VOID_ACTION'

handler<AppState, number>(NON_VOID_ACTION, (state, _payload) => state), // ok - state is AppState
handler<AppState, void>(VOID_ACTION, state => state) // ok - state is AppState

createReducer(
    defaultState,
    handler(NON_VOID_ACTION, (state, _payload) => state), // ok - state is AppState
    handler(VOID_ACTION, state => state) // unexpected - state is any
)
```

**Expected behavior:**
state parameter is infered to be "AppState" in the expression at the end
```ts
handler(VOID_ACTION, state => state)
```

**Actual behavior:**
state parameter is infered to be "any" 

**Playground Link:**  
[Playground](https://www.typescriptlang.org/play/index.html#src=%2F%2F%20Don't%20know%20what%20causes%20the%20problem%20so%20cannot%20come%20up%20with%20a%20shorter%20sample%0D%0A%0D%0Atype%20ActionType%3CP%3E%20%3D%20string%20%26%20%7B%20attachPayloadTypeHack%3F%3A%20P%20%26%20never%20%7D%0D%0A%0D%0A%2F*%0D%0A%2F%2F%20irrelevant%20type%20to%20reproduce%20my%20issue%0D%0A%2F%2F%20but%20gives%20context%20that%20action%20might%20have%20no%20additional%20payload%20to%20handle%0D%0A%0D%0Atype%20Action%3CP%3E%20%3D%20P%20extends%20void%0D%0A%20%20%20%20%3F%20%7B%20type%3A%20ActionType%3CP%3E%20%7D%0D%0A%20%20%20%20%3A%20%7B%20type%3A%20ActionType%3CP%3E%2C%20payload%3A%20P%20%7D%0D%0A*%2F%0D%0A%0D%0Atype%20Handler%3CS%2C%20P%3E%20%3D%20P%20extends%20void%0D%0A%20%20%20%20%3F%20(state%3A%20S)%20%3D%3E%20S%0D%0A%20%20%20%20%3A%20(state%3A%20S%2C%20payload%3A%20P)%20%3D%3E%20S%0D%0A%0D%0Ainterface%20ActionHandler%3CS%2C%20P%3E%20%7B%0D%0A%20%20%20%20actionType%3A%20ActionType%3CP%3E%0D%0A%20%20%20%20handler%3A%20Handler%3CS%2C%20P%3E%0D%0A%7D%0D%0A%0D%0Adeclare%20function%20handler%3CS%2C%20P%3E(actionType%3A%20ActionType%3CP%3E%2C%20handler%3A%20Handler%3CS%2C%20P%3E)%3A%20ActionHandler%3CS%2C%20P%3E%0D%0A%0D%0Adeclare%20function%20createReducer%3CS%3E(%0D%0A%20%20%20%20%20%20%20%20defaultState%3A%20S%2C%0D%0A%20%20%20%20%20%20%20%20...actionHandlers%3A%20ActionHandler%3CS%20%2F*%20force%20state%20S%20for%20every%20action%20handler*%2F%2C%20any%3E%5B%5D%0D%0A%20%20%20%20)%3A%20any%0D%0A%0D%0Ainterface%20AppState%20%7B%0D%0A%20%20%20%20dummy%3A%20string%0D%0A%7D%0D%0A%0D%0Aconst%20defaultState%3A%20AppState%20%3D%20%7B%0D%0A%20%20%20%20dummy%3A%20''%0D%0A%7D%0D%0A%0D%0Aconst%20NON_VOID_ACTION%3A%20ActionType%3Cnumber%3E%20%3D%20'NON_VOID_ACTION'%0D%0A%20%20%20%20%2C%20VOID_ACTION%3A%20ActionType%3Cvoid%3E%20%3D%20'VOID_ACTION'%0D%0A%0D%0Ahandler%3CAppState%2C%20number%3E(NON_VOID_ACTION%2C%20(state%2C%20_payload)%20%3D%3E%20state)%2C%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0Ahandler%3CAppState%2C%20void%3E(VOID_ACTION%2C%20state%20%3D%3E%20state)%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0A%0D%0AcreateReducer(%0D%0A%20%20%20%20defaultState%2C%0D%0A%20%20%20%20handler(NON_VOID_ACTION%2C%20(state%2C%20_payload)%20%3D%3E%20state)%2C%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0A%20%20%20%20handler(VOID_ACTION%2C%20state%20%3D%3E%20state)%20%2F%2F%20unexpected%20-%20state%20is%20any%0D%0A))
(use noImplicitAny flag)

**Related Issues:** 
Not sure #25301

**Workaround:**
When Handler is defined as follows
```ts
type Handler<S, P> = (state: S, payload: P) => S
```
it still allows to pass single argument handler and state is infered to be AppState

### LINKED ISSUE #25759: Function parameter type inference doesn't work with keyof of generic type
**TypeScript Version:**  3.1.0-dev.201xxxxx

**Search Terms:**
function parameter type inference keyof generic
**Code**

```ts
function foo1<T>(foo: T, bar: {[key in keyof {foo: number}] : (x: T) => number }) {
}

function foo2<T>(foo: T, bar: {[key in keyof T] : (x: T) => number }) {
}

foo1({foo: 1}, {foo: x1 => x1.foo + 1})
foo2({foo: 1}, {foo: x2 => x2.foo + 1})

```

**Expected behavior:**
`x2` has type `{foo:number}`
**Actual behavior:**
Compiler error 
`TS7006: Parameter 'x2' implicitly has an 'any' type.`

`x1` has type `{foo:number}` 
but
`x2` has type `any`

**Playground Link:** 
[Playground](http://www.typescriptlang.org/play/#src=function%20foo1%3CT%3E(foo%3A%20T%2C%20qwe%3A%20%7B%5Bkey%20in%20keyof%20%7Bfoo%3A%20number%7D%5D%20%3A%20(x%3A%20T)%20%3D%3E%20number%20%7D)%20%3A%20any%20%7B%0D%0A%20%20return%20%7B%7D%20as%20any%0D%0A%7D%0D%0A%0D%0Afunction%20foo2%3CT%3E(foo%3A%20T%2C%20qwe%3A%20%7B%5Bkey%20in%20keyof%20T%5D%20%3A%20(x%3A%20T)%20%3D%3E%20number%20%7D)%20%3A%20any%20%7B%0D%0A%20%20return%20%7B%7D%20as%20any%0D%0A%7D%0D%0A%0D%0Aconst%20f1%20%3D%20foo1(%7Bfoo%3A%201%7D%2C%20%7Bfoo%3A%20x%20%3D%3E%20x.foo%20%2B%201%7D)%0D%0Aconst%20f2%20%3D%20foo2(%7Bfoo%3A%201%7D%2C%20%7Bfoo%3A%20x%20%3D%3E%20x.foo%20%2B%201%7D)%0D%0A)

### LINKED ISSUE #25814: Contextual type inference fails in surprising case
**TypeScript Version:**  3.1.0-dev.20180719
**Code**

```ts
type R = {
  a: (x: number) => void;
  b: (x: string) => void;
};

type O = {
  on<P extends keyof R>(x: P, callback: R[P]): void;
};

declare var x: O;
x.on('a', a => {}); // not contextually typed.
```
**Expected behavior:**
Contextually type a to number.

**Actual behavior:**
A is not contextually typed, while signature help gives the correct answer.

**Playground Link:** [here](http://www.typescriptlang.org/play/index.html#src=type%20R%20=%20{%0D%0A%20%20a:%20(x:%20number)=%3Evoid%0D%0A%20%20b:%20(x:%20string)=%3Evoid%0D%0A}%0D%0Atype%20O%20=%20{%0D%0A%20%20on%3CP%20extends%20keyof%20R%3E(x:%20P,%20cb:%20R[P]):%20void;%0D%0A}%0D%0Adeclare%20var%20x:%20O%0D%0Ax.on('a',%20a=%3E{})%20//%20why%20no%20contextual%20typing?%20signature%20help%20gives%20correct%20answer)

### LINKED ISSUE #29775: Type inference fails to apply to functions as args
**TypeScript Version:**  3.2 and up (possibly older ones too)

**Code**

```ts
declare class Component<P> { 
  constructor(props: P);
}
    
interface ComponentClass<P = {}> {
  new (props: P): Component<P>;
}

type CreateElementChildren<P> =
  P extends { children?: infer C }
  ? C extends any[]
    ? C
    : C[]
  : unknown;

declare function createElement<P extends {}>(
  type: ComponentClass<P>,
  ...children: CreateElementChildren<P>
): any;

declare function createElement2<P extends {}>(
  type: ComponentClass<P>,
  child: CreateElementChildren<P>
): any;
        
// To be sure the optional isn't the problem:
class InferFunctionTypes extends Component<{children: (foo: number) => string}> {}
createElement(InferFunctionTypes, (foo) => "" + foo); // ERROR! Parameter 'foo' implicitly has an 'any' type.
// In the above, P is properly infered on the createElement, per hover tooltip:
// function createElement<{ children: (foo: number) => string; }>(type: ComponentClass<{ children: (foo: number) => string; }>, ...children: ((foo: number) => string)[]): any 

// If you don't expect P to be infered, it works properly:
createElement<{ children: (foo: number) => string }>(InferFunctionTypes, foo => "" + foo);

// Is unrelated to varargs
createElement2(InferFunctionTypes, [(foo) => "" + foo]); // ERROR! Parameter 'foo' implicitly has an 'any' type.

```

**Expected behavior:**

Typescript should be able to infer that `foo` is a `number`.

**Actual behavior:**

`Parameter 'foo' implicitly has an 'any' type.`

**Playground Link:** 
http://www.typescriptlang.org/play/#src=declare%20class%20Component%3CP%3E%20%7B%20%0A%20%20constructor(props%3A%20P)%3B%0A%7D%0A%20%20%20%20%0Ainterface%20ComponentClass%3CP%20%3D%20%7B%7D%3E%20%7B%0A%20%20new%20(props%3A%20P)%3A%20Component%3CP%3E%3B%0A%7D%0A%0Atype%20CreateElementChildren%3CP%3E%20%3D%0A%20%20P%20extends%20%7B%20children%3F%3A%20infer%20C%20%7D%0A%20%20%3F%20C%20extends%20any%5B%5D%0A%20%20%20%20%3F%20C%0A%20%20%20%20%3A%20C%5B%5D%0A%20%20%3A%20unknown%3B%0A%0Adeclare%20function%20createElement%3CP%20extends%20%7B%7D%3E(%0A%20%20type%3A%20ComponentClass%3CP%3E%2C%0A%20%20...children%3A%20CreateElementChildren%3CP%3E%0A)%3A%20any%3B%0A%0Adeclare%20function%20createElement2%3CP%20extends%20%7B%7D%3E(%0A%20%20type%3A%20ComponentClass%3CP%3E%2C%0A%20%20child%3A%20CreateElementChildren%3CP%3E%0A)%3A%20any%3B%0A%20%20%20%20%20%20%20%20%0A%2F%2F%20To%20be%20sure%20the%20optional%20isn't%20the%20problem%3A%0Aclass%20InferFunctionTypes%20extends%20Component%3C%7Bchildren%3A%20(foo%3A%20number)%20%3D%3E%20string%7D%3E%20%7B%7D%0AcreateElement(InferFunctionTypes%2C%20(foo)%20%3D%3E%20%22%22%20%2B%20foo)%3B%20%2F%2F%20ERROR!%20Parameter%20'foo'%20implicitly%20has%20an%20'any'%20type.%0A%2F%2F%20In%20the%20above%2C%20P%20is%20properly%20infered%20on%20the%20createElement%2C%20per%20hover%20tooltip%3A%0A%2F%2F%20function%20createElement%3C%7B%20children%3A%20(foo%3A%20number)%20%3D%3E%20string%3B%20%7D%3E(type%3A%20ComponentClass%3C%7B%20children%3A%20(foo%3A%20number)%20%3D%3E%20string%3B%20%7D%3E%2C%20...children%3A%20((foo%3A%20number)%20%3D%3E%20string)%5B%5D)%3A%20any%20%0A%0A%2F%2F%20If%20you%20don't%20expect%20P%20to%20be%20infered%2C%20it%20works%20properly%3A%0AcreateElement%3C%7B%20children%3A%20(foo%3A%20number)%20%3D%3E%20string%20%7D%3E(InferFunctionTypes%2C%20foo%20%3D%3E%20%22%22%20%2B%20foo)%3B%0A%0A%2F%2F%20Is%20unrelated%20to%20varargs%0AcreateElement2(InferFunctionTypes%2C%20%5B(foo)%20%3D%3E%20%22%22%20%2B%20foo%5D)%3B%0A

**Related Issues:**

### LINKED ISSUE #30341: Incorrect type inference when conditionally inferring multiple types
**TypeScript Version:**  3.4.0-dev.20190312

**Code**

The following code throws the indicated error:

```ts
type Action<T extends string, P> = { readonly type: T; payload: P; }
type ActionCreator<T extends string, P> = { readonly type: T, (payload: P): Action<T, P> };
type Reducer<S, T extends string, P> = (state: S, action: Action<T, P>) => S;
type ReducerFor<S, AC> = AC extends ActionCreator<infer T, infer P> ? Reducer<S, T, P> : never;

declare function handleAction<AC>(action: AC, reducer: ReducerFor<{ value: string }, AC>): void;
declare const SomeAction: ActionCreator<"TestAction", string>;

handleAction(SomeAction, (state, action) => ({ ...state, value: action.payload }));
/*                               ^ resolved as Action<string, {}>      ^ error appears here:
Type '{}' is not assignable to type 'string'.ts(2322)
The expected type comes from property 'value' which is declared here on type '{ value: string; }'
*/
```
As indicated in the comments, `action` is being resolved as `Action<string, {}>` instead of `Action<"TestAction", string>`.

However, if I write the conditional inference in a different way, then this will work:

```ts
type ReducerFor<S, AC> = 
   Reducer<S, AC extends ActionCreator<infer T, any> ? T : never, 
              AC extends ActionCreator<any, infer P> ? P : never>;

declare const SomeAction: ActionCreator<"TestAction", string>;

handleActionTest(SomeAction, (state, action) => ({ ...state, value: action.payload }));
//                                   ^ correctly resolved as Action<"TestAction", string>
```

**Expected behavior:**
The compiler should correctly handle the type inference and return a type of `Action<"TestAction", string>`

**Actual behavior:**
Incorrect type inference to `Action<string, {}>`

**Playground Link:** 
[Link with issue](https://www.typescriptlang.org/play/index.html#src=type%20Action%3CT%20extends%20string%2C%20P%3E%20%3D%20%7B%0D%0A%20%20%20%20readonly%20type%3A%20T%0D%0A%20%20%20%20payload%3A%20P%3B%0D%0A%7D%0D%0A%0D%0Atype%20ActionCreator%3CT%20extends%20string%2C%20P%3E%20%3D%20%7B%20%0D%0A%20%20%20%20readonly%20type%3A%20T%2C%20%0D%0A%20%20%20%20(payload%3A%20P)%3A%20Action%3CT%2C%20P%3E%20%0D%0A%7D%3B%0D%0A%0D%0Atype%20Reducer%3CS%2C%20T%20extends%20string%2C%20P%3E%20%3D%20(state%3A%20S%2C%20action%3A%20Action%3CT%2C%20P%3E)%20%3D%3E%20S%3B%0D%0A%0D%0Atype%20ReducerFor%3CS%2C%20AC%3E%20%3D%20AC%20extends%20ActionCreator%3Cinfer%20T%2C%20infer%20P%3E%20%3F%20Reducer%3CS%2C%20T%2C%20P%3E%20%3A%20never%3B%0D%0A%0D%0Adeclare%20const%20SomeAction%3A%20ActionCreator%3C%22TestAction%22%2C%20string%3E%3B%0D%0Adeclare%20function%20handleActionTest%3CAC%3E(action%3A%20AC%2C%20reducer%3A%20ReducerFor%3C%7B%20value%3A%20string%20%7D%2C%20AC%3E)%3A%20void%3B%0D%0A%0D%0AhandleActionTest(SomeAction%2C%20(state%2C%20action)%20%3D%3E%20(%7B%20...state%2C%20value%3A%20action.payload%20%7D))%3B)

[Link with workaround](https://www.typescriptlang.org/play/index.html#src=type%20Action%3CT%20extends%20string%2C%20P%3E%20%3D%20%7B%0D%0A%20%20%20%20readonly%20type%3A%20T%0D%0A%20%20%20%20payload%3A%20P%3B%0D%0A%7D%0D%0A%0D%0Atype%20ActionCreator%3CT%20extends%20string%2C%20P%3E%20%3D%20%7B%20%0D%0A%20%20%20%20readonly%20type%3A%20T%2C%20%0D%0A%20%20%20%20(payload%3A%20P)%3A%20Action%3CT%2C%20P%3E%20%0D%0A%7D%3B%0D%0A%0D%0Atype%20Reducer%3CS%2C%20T%20extends%20string%2C%20P%3E%20%3D%20(state%3A%20S%2C%20action%3A%20Action%3CT%2C%20P%3E)%20%3D%3E%20S%3B%0D%0A%0D%0Atype%20ReducerFor%3CS%2C%20AC%3E%20%3D%20Reducer%3CS%2C%20AC%20extends%20ActionCreator%3Cinfer%20T%2C%20any%3E%20%3F%20T%20%3A%20never%2C%20AC%20extends%20ActionCreator%3Cany%2C%20infer%20P%3E%20%3F%20P%20%3A%20never%3E%3B%0D%0A%0D%0Adeclare%20const%20SomeAction%3A%20ActionCreator%3C%22TestAction%22%2C%20string%3E%3B%0D%0Adeclare%20function%20handleActionTest%3CAC%3E(action%3A%20AC%2C%20reducer%3A%20ReducerFor%3C%7B%20value%3A%20string%20%7D%2C%20AC%3E)%3A%20void%3B%0D%0A%0D%0AhandleActionTest(SomeAction%2C%20(state%2C%20action)%20%3D%3E%20(%7B%20...state%2C%20value%3A%20action.payload%20%7D))%3B)

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we improve our ability to contextually type parameters of arrow functions and function expressions passed as arguments, allowing contextual types to depend on inferences made from other arguments. For example:

```ts
type EventMap = {
  a: () => void,
  b: (x: number, y: number) => number;
  c: (s: string) => boolean;
}

type EventObject = {
  on<K extends keyof EventMap>(key: K, handler: EventMap[K]): ReturnType<EventMap[K]>;
}

declare var obj: EventObject;
obj.on('a', () => {});            // () => void
obj.on('b', (x, y) => x * y);     // (x: number, y: number) => number
obj.on('c', s => s.length > 10);  // (s: string) => boolean
```

Above, the contextual type for each arrow function arguments depends on the type inferred for `K` from the first argument. Specifically, the contextual type for the arrow function argument is instantiated with inferences made from other arguments before the contextual type is applied to the arrow function parameters.

Fixes #6611.
Fixes #21382.
Fixes #22149.
Fixes #25299.
Fixes #25759.
Fixes #25814.
Fixes #29775.
Fixes #30341.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this.

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
The RWC tests are clean. The DT tests have a few changes that are to be expected: Now, when type inference produces no candidates and a type parameter has no explicit default, we consistently default to the constraint of the type parameter. Previously we'd default to `{}` if `{}` was assignable to the constraint. For example, we'd default to `{}` when the constraint is `object`.

**Comment by DanielRosenwasser** [comment]
> ```ts
> type EventMap = {
>   a: () => void,
>   b: (x: number, y: number) => number;
>   c: (s: string) => boolean;
> }
> 
> type EventObject = {
>   on<K extends keyof EventMap>(key: K, handler: EventMap[K]): ReturnType<EventMap[K]>;
> }
> 
> declare var obj: EventObject;
> obj.on('a', () => {});            // () => void
> obj.on('b', (x, y) => x * y);     // (x: number, y: number) => number
> obj.on('c', s => s.length > 10);  // (s: string) => boolean
> ```

The only example that doesn't work right now is the call to `obj.on('c', ...)`, where you'll get an implicit `any` error:

````
Parameter 's' implicitly has an 'any' type.
````

Why does the call to `on('b', ...)` work as desired?

**Comment by ahejlsberg** [comment]
> Why does the call to on('b', ...) work as desired?

We do some rudimentary signature filtering based on arity and since there is only one signature with arity two or greater we were able to produce a contextual signature for the `b` case.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20258,7 +20264,7 @@ namespace ts {
                         getArrayifiedType(checkExpressionWithContextualType((<SpreadElement>arg).expression, restType, context, CheckMode.Normal));
                 }
             }
-            const contextualType = getIndexTypeOfType(restType, IndexKind.Number) || anyType;
+            const contextualType = getIndexedAccessType(restType, numberType);
```
Comment (weswigham) [inline src/compiler/checker.ts]:
```suggestion
            const contextualType = getIndexedAccessType(restType, getLiteralType(index));
```
?

Comment (ahejlsberg) [reply src/compiler/checker.ts]: In principle yes, but not exactly like that. Just pushed a new commit.

**Pull request merged**